test/spark/app/build.sbt

```diff
@@ -1,11 +1,69 @@
-name := "sonnets"
+import build.BuildType
 
-version := "0.1"
+val baseName = "sonnets"
 
-scalaVersion := "2.12.10"
+val projectVersion = "0.1.0"
 
-libraryDependencies ++= Seq(
-    "org.apache.spark" %% "spark-sql" % "3.1.1" % "provided",
-    "com.amazonaws" % "aws-java-sdk-bundle" % "1.11.375" % "provided",
-    "org.apache.hadoop" % "hadoop-aws" % "3.2.0" % "provided",
+// Spark versions 2.4.7 and 3.0.1 use different Scala versions.  Changing this is a deep
+// change, so key the Spark distinction by the Scala distinction.  sbt doesn't appear to
+// support other ways of changing emitted Scala binary versions using the same compiler.
+
+// SO https://stackoverflow.com/a/60177627/192263 hints that we cannot use 2.11 here before
+// this version
+val scala211Version = "2.11.12"
+val scala212Version = "2.12.13"
+
+def settingsToCompileIn() = {
+  Seq(
+    Compile / scalaSource := (ThisBuild / baseDirectory).value / "src" / "main" / "scala",
+  )
+}
+
+def generateProject(buildType: BuildType) =
+  Project(s"${baseName}-${buildType.name}", file(s"target/${baseName}-${buildType.name}"))
+    .settings(
+      sharedSettings,
+      settingsToCompileIn(),
+      scalaVersion := buildType.scalaVersion,
+      libraryDependencies ++= Seq(
+        "org.apache.spark" %% "spark-sql" % buildType.sparkVersion % "provided",
+        "org.apache.hadoop" % "hadoop-aws" % buildType.hadoopVersion,
+        "org.apache.hadoop" % "hadoop-common" % buildType.hadoopVersion,
+      ),
+      target := { baseDirectory.value / "target" / s"${baseName}-${buildType.name}" }
+    )
+
+lazy val proj24 = generateProject(new BuildType("246", scala211Version, "2.4.6", "2.7.7"))
+lazy val proj31 = generateProject(new BuildType("311", scala212Version, "3.1.1", "2.7.7"))
+
+lazy val root = (project in file("."))
+  .aggregate(proj24, proj31)
+  .settings(
+      compile / skip := true,
+      assembly / skip := true,
+      publish / skip := true,
+  )
+
+
+lazy val assemblySettings = Seq(
+  assembly / assemblyMergeStrategy := (_ => MergeStrategy.first),
+  assembly / assemblyShadeRules := Seq(
+    ShadeRule.rename("scala.collection.compat.**" -> "shadecompat.@1").inAll,
```

Comment:
Strange that we need to shade _this_: I would expect it to be a part of the Scala runtime.

Reply:
right. I'll remove all shading

Reply:
and assembly